### PR TITLE
JsonMenuEditorField 

### PR DESCRIPTION
**Pull request by theus77**

### CODE REVIEW

File: Resources/assets/edit-revision.js [hunk 1]
```diff
@@ -454,10 +454,9 @@ function addEventListeners(target){
         onFormChange();
     });
 
-    target.find("input").keypress(onFormChange);
-    target.find("input").change(onFormChange);
-    target.find("select").change(onFormChange);
-    target.find("textarea").keypress(onFormChange);
+    target.find("input").on('input', onFormChange);
+    target.find("select").on('change', onFormChange);
+    target.find("textarea").on('input', onFormChange);
```
Comment (dameert) [inline Resources/assets/edit-revision.js]: No longer “keypress” and one got removed?

File: Resources/views/macros/data-field-type.html.twig [hunk 2]
```diff
@@ -637,6 +637,35 @@
 	{{ _self.messages(dataField) }}
 {% endblock %}
 
+
+
+{% block EMS_CoreBundle_Form_DataField_JsonMenuEditorFieldType %}
+	{% set doCompare = compare and dataField.rawData is not same as( compareRawData|get_string(dataField.fieldType.name) )  %}
```
Comment (dameert) [inline Resources/views/macros/data-field-type.html.twig]: DoCompare seems business logic for PhP side.

File: Resources/views/macros/data-field-type.html.twig [hunk 3]
```diff
@@ -637,6 +637,35 @@
...
+	{% set color = not doCompare ? '' : not dataField.rawData ? 'bg-red' : not compareRawData|get_string(dataField.fieldType.name) ? 'bg-green' : 'bg-orange' %}
```
Comment (dameert) [inline Resources/views/macros/data-field-type.html.twig]: Color seems business logic for php side. Okay color is frontend. But that is css issue. Determining the class that has to be used is a business case, for PHP

File: Form/DataField/JsonMenuEditorFieldType.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        return HiddenType::class;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        $fieldType = $options['metadata'];
```
Comment (dameert) [inline Form/DataField/JsonMenuEditorFieldType.php]: Fieldtype is only used in line 59. Move this line to line 58? And add a /** @var for IDE help?

File: Form/DataField/JsonMenuEditorFieldType.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+        /*get options for twig context*/
```
Comment (dameert) [inline Form/DataField/JsonMenuEditorFieldType.php]: Comment not needed

File: Form/DataField/JsonMenuEditorFieldType.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        parent::buildView($view, $form, $options);
+        $view->vars ['icon'] = $options ['icon'];
```
Comment (dameert) [inline Form/DataField/JsonMenuEditorFieldType.php]: Why not build the $attr first. Then assign $view->vars = \array_merge($view->vars, $attr) ? This way you only have one line setting data in this public variable (easier to maintain if one day this vars property becomes private for php.

Comment (dameert) [reply Form/DataField/JsonMenuEditorFieldType.php]: Oops should be \array_merge($view->vars, [‘attr’ => $attr, ‘icon’ => $options[‘icon’]])

File: Form/DataField/JsonMenuEditorFieldType.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        if (empty($attr['class'])) {
+            $attr['class'] = '';
+        }
+
+        $attr['data-locales'] = $options['locales'];
+        $attr['data-maxDepth'] = $options['maxDepth'];
+        $attr['data-disabled'] = !$this->authorizationChecker->isGranted($fieldType->getMinimumRole());
+        $attr['class'] .= ' code_editor_ems';
+
+        $view->vars ['attr'] = $attr;
```
Comment (dameert) [inline Form/DataField/JsonMenuEditorFieldType.php]: Seeing many spaces in code before arrays. Phpcs?

File: Form/DataField/JsonMenuEditorFieldType.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        return 'json_menu_editor_fieldtype';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        /*set the default option value for this kind of compound field*/
```
Comment (dameert) [inline Form/DataField/JsonMenuEditorFieldType.php]: The code is allready saying this. Comment not needed.

File: Form/DataField/JsonMenuEditorFieldType.php [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+     *
+     * {@inheritdoc}
+     *
+     */
+    public function buildOptionsForm(FormBuilderInterface $builder, array $options)
+    {
+        parent::buildOptionsForm($builder, $options);
+        $optionsForm = $builder->get('options');
+
+        // String specific mapping options
```
Comment (dameert) [inline Form/DataField/JsonMenuEditorFieldType.php]: I don’t understand the comment. But the code is clear. Comment can be removed.

File: Resources/assets/edit-revision.js (hunk 1 above)
Comment (theus77) [reply Resources/assets/edit-revision.js]: input event is much more efficient than the keypress event

File: Resources/views/macros/data-field-type.html.twig (hunk 2 above)
Comment (theus77) [reply Resources/views/macros/data-field-type.html.twig]: For once I disagree, It's the view to compare fields between two revision

File: Resources/views/macros/data-field-type.html.twig (hunk 3 above)
Comment (theus77) [reply Resources/views/macros/data-field-type.html.twig]: I' dont see how we can handle this.

File: Form/DataField/JsonMenuEditorFieldType.php (hunk 9 above)
Comment (theus77) [reply Form/DataField/JsonMenuEditorFieldType.php]: Fixed

File: Form/DataField/JsonMenuEditorFieldType.php (hunk 8 above)
Comment (theus77) [reply Form/DataField/JsonMenuEditorFieldType.php]: fixed

File: Form/DataField/JsonMenuEditorFieldType.php (hunk 7 above)
Comment (theus77) [reply Form/DataField/JsonMenuEditorFieldType.php]: Done

File: Form/DataField/JsonMenuEditorFieldType.php (hunk 6 above)
Comment (theus77) [reply Form/DataField/JsonMenuEditorFieldType.php]: Fixed

File: Form/DataField/JsonMenuEditorFieldType.php (hunk 5 above)
Comment (theus77) [reply Form/DataField/JsonMenuEditorFieldType.php]: Ok, done

File: Form/DataField/JsonMenuEditorFieldType.php (hunk 4 above)
Comment (theus77) [reply Form/DataField/JsonMenuEditorFieldType.php]: Ok, done

File: Resources/views/macros/data-field-type.html.twig (hunk 3 above)
Comment (theus77) [reply Resources/views/macros/data-field-type.html.twig]: Maybe extend the FieldType interface (renderCompare ?) acting like a mini controller? (another pull request)

File: Form/DataField/JsonMenuEditorFieldType.php (hunk 7 above)
Comment (dameert) [reply Form/DataField/JsonMenuEditorFieldType.php]: nope still seeing spaces between the property and the array key

Comment (theus77) [reply Form/DataField/JsonMenuEditorFieldType.php]: Done

**Pull request merged**